delivery_price_by_category/models/delivery_carrier.py

```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Simone Rubino - Agile Business Group
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, models
+from odoo.tools import safe_eval
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+
+    @api.multi
+    def get_price_available(self, order):
+        if 'order_id' not in self.env.context:
+            return super(DeliveryCarrier, self) \
+                .with_context(order_id=order.id).get_price_available(order)
+        return super(DeliveryCarrier, self).get_price_available(order)
+
+    def get_price_from_picking(self, total, weight, volume, quantity):
```

Comment:
@SimoRubi I think your assertion is not correct. I would have done your tests in get_price_available function as the only condition to compute in the other manner is line.product_category_id. Don't override get_price_from_picking. And then you avoid annoying context changes.

Reply:
Thanks I'll work on your approach, that might actually simplify the whole process

Reply:
@rousseldenis I tried your approach in https://github.com/OCA/delivery-carrier/pull/165/commits/5005b6dcd6dcd48b0b8a0a3318bb23a36760b580, is that what you meant?